File: day7/day7.py
class Bag():
    def __init__(self, name, holds):
        self.name = name
        self.holds = holds

def fill_bags_dict(filename):
    bags_dict = {}
    with open(filename) as file:
        for line in file:
            contain_split = line.replace(".", "").replace("\n", "").split('contain')
            capacity_split = contain_split[1].split(',')
            bag_name = contain_split[0][:-1]
            if "no other bags" in capacity_split[0]:
                bags_dict[bag_name] = Bag(bag_name, [])
                continue
            holds = list()
            for c in capacity_split:
                holds.append([int(c[1]), c[3:]])
            bags_dict[bag_name] = Bag(bag_name, holds)
    return bags_dict

def can_hold_shiny_bag(key, bags_dict):
    for bag in bags_dict[key].holds:
        if "shiny gold" in bag[1]:
            return True
        if can_hold_shiny_bag(bag[1], bags_dict):
            return True
    return False

def bags_in_bag(key, bags_dict):
    count = 0
    for bag in bags_dict[key].holds:
        if bag[0]:
            count += bag[0] + bag[0] * bags_in_bag(bag[1], bags_dict)
    return count

def main():
    answer_p1 = 0
    bags_dict = fill_bags_dict('day7.in')

    for key in bags_dict.keys():
        if can_hold_shiny_bag(key, bags_dict):
            answer_p1 += 1
    
    answer_p2 = bags_in_bag("shiny gold bags", bags_dict)
    print("p1:", answer_p1)
    print("p2:", answer_p2)

main()

File: day7/day7.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Holds struct {
	amount int
	name   string
}

type Bag struct {
	name  string
	holds []Holds
}

func contains(v string, slice []string) bool {
	for _, s := range slice {
		if v == s {
			return true
		}
	}
	return false
}

func fillBagsMap(filename string) map[string]Bag {
	bagsMap := make(map[string]Bag)

	file, _ := os.Open(filename)

	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		line := scanner.Text()
		line = strings.Replace(strings.Replace(line, ".", "", 1), "\n", "", 1)
		containSplit := strings.Split(line, "contain")
		capacitySplit := strings.Split(containSplit[1], ",")
		bagName := strings.TrimRight(containSplit[0], " ")
		holds := make([]Holds, 0)
		if contains("no other bags", capacitySplit) {
			holds = append(holds, Holds{amount: 0, name: "no other bags"})
			bag := Bag{name: bagName, holds: holds}
			bagsMap[bagName] = bag
			continue
		}
		for _, c := range capacitySplit {
			amount, _ := strconv.Atoi(string(c[1]))
			name := c[3:]
			holds = append(holds, Holds{amount: amount, name: name})
		}
		bag := Bag{name: bagName, holds: holds}
		bagsMap[bagName] = bag
	}

	return bagsMap
}

func holdsShinyBag(key string, bagsMap map[string]Bag) bool {
	for _, bag := range bagsMap[key].holds {
		if "shiny gold bags" == bag.name {
			return true
		}
		if holdsShinyBag(bag.name, bagsMap) {
			return true
		}
	}
	return false
}

func bagsInBag(key string, bagsMap map[string]Bag) int {
	count := 0
	for _, bag := range bagsMap[key].holds {
		if bag.amount > 0 {
			count += bag.amount + bag.amount*bagsInBag(bag.name, bagsMap)
		}
	}
	return count
}

func main() {
	answerP1 := 0
	bagsMap := fillBagsMap("day7.in")
	for key := range bagsMap {
		if holdsShinyBag(key, bagsMap) {
			answerP1++
		}
	}
	answerP2 := bagsInBag("shiny gold bags", bagsMap)
	fmt.Println(answerP1)
	fmt.Println(answerP2)
}
